File: 589. N-ary Tree Preorder Traversal/Solution.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val, children):
        self.val = val
        self.children = children
"""
class Solution(object):
    def preorder(self, root):
        """
        :type root: Node
        :rtype: List[int]
        """
        result = []
        self._visit(result, root)
        return result
    
    def _visit(self, result, root):
        if root is not None:
            result.append(root.val)
            for ch in root.children:
                self._visit(result, ch)

File: 589. N-ary Tree Preorder Traversal/Solution.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
public:
    vector<int> preorder(Node* root) {
        vector<int> result;
        visit(result, root);
        return result;
        
    }
    
    void visit(vector<int> & res, Node * node) {
        if (node) {
            res.push_back(node->val);
            for (auto & ch : node->children) visit(res, ch);
        }
    }
};

